crates/network/README.md

```diff
@@ -0,0 +1,302 @@
+
+#### Main Loop
+
+The EventLoop's main loop uses the `tokio::select!` macro that allows waiting on multiple async computations and returns when a single computation completes :
+
+1. **Swarm Events**:
+   Network-related events from libp2p, such as new connections, disconnections, and protocol-specific events. These events are core to managing the peer-to-peer network state.
+
+2. **Incoming Streams**:
+   New incoming stream connections from other peers. This allows the node to handle requests and communications initiated by other nodes in the network.
+
+3. **Commands**:
+   Instructions received from other parts of the application through a command channel. This enables other components to interact with and control the network operations.
+
+4. **Periodic Rendezvous Discoveries**:
+   Regular attempts to discover new peers through the Rendezvous protocol. This helps maintain and expand the node's network of peers.
+
+Each of the events has it's own handler function.
+
+### Swarm
+
+The Swarm is a central construct in libp2p, providing a view of the network from the perspective of a single node. It manages connections, protocols, and peer interactions.   In our implementation, the Swarm is defined as:
+
+ ```rust
+  Swarm<Behaviour>
+```
+
+Where  `Behaviour`  is a custom type that implements the  `NetworkBehaviour`  trait. This allows us to define specific network behaviors tailored to our application's needs.
+
+The Swarm in libp2p combines a `Transport` (how to send/receive data) with a `NetworkBehaviour` (what to do with the connections).
+
+### Identity and Peer ID
+
+Identity in libp2p is crucial for unique node identification and secure communications. Each node has a Peer ID derived from its cryptographic key pair. This Peer ID is used in multiaddresses, allowing for persistent addressing:
+
+    /ip4/192.0.2.0/tcp/443/p2p/QmcEPrat8ShnCph8WjkREzt5CPXF2RwhYxYBALDcLC1iV6
+The Peer ID (after `/p2p/`) enables connections to persist even if IP addresses change, enhancing network resilience.
+
+### Runtime
+
+This networking crate utilizes Tokio as its asynchronous runtime, leveraging its efficient task scheduler and I/O operations to handle concurrent network activities and event processing.
+
+### Transport Protocols
+
+The networking crate supports multiple transport protocols to ensure wide compatibility and optimal performance:
+
+- TCP: Reliable, connection-oriented protocol for general-purpose communication.
+- QUIC: A modern, multiplexed transport built on UDP, offering improved performance and security.
+
+These protocols are configured with encryption (TLS for QUIC, Noise for TCP) and multiplexing (QUIC's native multiplexing, Yamux for TCP) to ensure secure and efficient data transfer.
```

Comment:
I think this is TLS, otherwise Noise for TCP, and QUIC is defined to be TLS non-optionally